### DISCUSSION TITLE
How to get the area and volum of IFCSPACE instance by using the data of Geometrical Representation or other way?

### DISCUSSION THREAD

**Discussion by u0416033** [IfcOpenShell/IfcOpenShell#2862] [Q&A]
How to get the area and volum of IFCSPACE instance by using the data of Geometrical Representation or other way?
I try to get some information of the IFCSPACE instance , and I have try some way but always lack of some properties.
`import ifcopenshell

ifc_file = ifcopenshell.open("your_file.ifc")
spaces = ifc_file.by_type("IfcSpace")

for space in spaces:
    if space.Representation:
        for rep in space.Representation.Representations:
            if rep.RepresentationIdentifier == "Body":
                if rep.RepresentationType == "SweptSolid":
                    volume = rep.Items[0].SweptArea.OuterCurve.Area * rep.Items[0].SweptArea.Position.Depth
                    print("Volume of", space.Name, "is", volume, "cubic meters")`
Like this .I try to get the Geometrical Representation data of the elemet but it was failed. 
How can I extract the data from Geometrical Representation or other way to get the coordinate of the SweptSolid of IFCSPACE?

![image](https://user-images.githubusercontent.com/53708536/224656466-b8738d7f-bd22-4c3d-b9ce-26e936898758.png)

**Comment by Moult** [comment]
See ifcopenshell.util.shape. It has functions to calculate areas, footprints, perimeters, volume, etc.

**Reply by u0416033** [reply]
I see ,thank you!  But I think I need to get the vertices of the room if it is a polygon. Are there any modules can do that?

**Reply by u0416033** [reply]
`settings = ifcopenshell.geom.settings()

shape = ifcopenshell.geom.create_shape(settings, space)

verts = shape.geometry.verts
face = shape.geometry.faces

grouped_verts = [[verts[i], verts[i + 1], verts[i + 2]] for i in range(0, len(verts), 3)]

print(grouped_verts)`

I have got the way to extract these! Thank you a lot.

**Reply by u0416033** [reply]
@Moult hi,I am curious about that I try to use the 
`ifcspace.get_info(recursive=True)['Representation'] `
to get the information about the room and it return a data like JSON but not a correct JSON format.
Can I have any way to deal with the return data like normal JSON format data?
This is the returned data below :
```
{"id": 471, "type": "IfcProductDefinitionShape", "Name": None, "Description": None, "Representations": ({"id": 469, "type": "IfcShapeRepresentation", "ContextOfItems": {"id": 117, "type": "IfcGeomontext": {"id": 112, "type": "IfcGeometricRepresentationContext", "ContextIdentifier": None, "ContextType": "Model", "CoordinateSpaceDimension": 3, "Precision": 0.001, "WorldCoordinateSystem": {"id": 109, "type": "IfcAxis2Placement3D", "Location": {"id": 6, "type": "IfcCartesianPoint", "Coordinates": (0.0, 0.0, 0.0)}, "Axis": None, "RefDirection": None}, "TrueNorth": {"id": 110, "type": "IfcDirection", "DirectionRatios": (6.12303176911189e-17, 1.0)}}, "TargetScale": None, "TargetView": "MODEL_VIEW", "UserDefinedTargetView": None}, "RepresentationIdentifier": "Body", "RepresentationType": "SweptSolid", "Items": ({"id": 468, "type": "IfcExtrudedAreaSolid", "SweptArea": {"id": 464, "type": "IfcArbitraryClosedProfileDef", "ProfileType": "AREA", "ProfileName": None, "OuterCurve": {"id": 462, "type": "IfcPolyline", "Points": ({"id": 438, "type": "IfcCartesianPoint", "Coordinates": (-255.823889594443, 57.2129170906261)}, {"id": 440, "type": "IfcCartesianPoint", "Coordinates": (-255.823889594443, 87.2129170906262)}, {"id": 442, "type": "IfcCartesianPoint", "Coordinates": (-495.420556261112, 87.2129170906269)}, {"id": 444, "type": "IfcCartesianPoint", "Coordinates": 
(-495.420556261112, -222.735312878815)}, {"id": 446, "type": "IfcCartesianPoint", "Coordinates": (524.286149140554, -222.735312878813)}, {"id": 448, "type": "IfcCartesianPoint", "Coordinates": (524.286149140554, -210.832671920612)}, {"id": 450, "type": "IfcCartesianPoint", "Coordinates": (241.391093154722, 78.5999609532321)}, {"id": 452, "type": "IfcCartesianPoint", "Coordinates": (241.391093154722, 57.2129170906253)}, {"id": 454, "type": "IfcCartesianPoint", "Coordinates": (191.391093154722, 57.2129170906253)}, {"id": 456, "type": "IfcCartesianPoint", "Coordinates": (191.391093154722, 87.2129170906256)}, {"id": 458, "type": "IfcCartesianPoint", "Coordinates": (-205.823889594443, 87.2129170906262)}, {"id": 460, "type": "IfcCartesianPoint", "Coordinates": (-205.823889594443, 57.2129170906261)}, {"id": 438, "type": "IfcCartesianPoint", "Coordinates": (-255.823889594443, 57.2129170906261)})}}, "Position": {"id": 467, "type": "IfcAxis2Placement3D", "Location": {"id": 465, "type": "IfcCartesianPoint", "Coordinates": (-1055.66481595393, -804.168725560609, 0.0)}, "Axis": {"id": 21, "type": "IfcDirection", "DirectionRatios": (0.0, 0.0, -1.0)}, "RefDirection": {"id": 17, "type": "IfcDirection", "DirectionRatios": (0.0, -1.0, 0.0)}}, "ExtrudedDirection": {"id": 21, "type": "IfcDirection", "DirectionRatios": (0.0, 0.0, -1.0)}, "Depth": 400.0},)},)}
```

**Reply by aothms** [reply]
What you're seeing is a representation of a nested Python dictionary, which looks like JSON by coincidence, but which is not strictly JSON (non-string keys are possible, distinction between list, tuples, sets, etc). To turn it into pure JSON you just need to do

```python
import json
print(json.dumps(ifcspace.get_info(recursive=True)['Representation']))
```

**Reply by u0416033** [reply]
@aothms Thank you! I have another question about the IFCSPACE "Global Z" value.I see the value in the BIMvision software, and I try to get the value by the code below:
`if ifcspace.ObjectPlacement.is_a('IfcLocalPlacement'):
    location = ifcspace.ObjectPlacement.RelativePlacement.Location
    x = location.Coordinates[0]
    y = location.Coordinates[1]
    z = location.Coordinates[2]` 
    
but it returned (0,0,0) . Is there any way to get the value -275?
![image](https://user-images.githubusercontent.com/53708536/225199028-fbba08f1-aa98-4cd2-acb8-46341247f391.png)

**Reply by Moult** [reply]
That sounds like a computed value. You can do it via:

```
    z_values = [geometry.verts[i + 2] for i in range(0, len(geometry.verts), 3)]
    top_elevation = max(z_values)
    bottom_elevation = min(z_values)
```

**Reply by u0416033** [reply]
@Moult  I try the code but it will return the vertices of SweptSolid ,so vertice will be start with the Z=0. But I try to get it real height of the IFCSPACE according to the IFCBuildingStorey. Or is there any way to get the storey of the IFCSPACE?
I have try it below but return null:

`if space.ContainsElements:
        # 取得與 IfcSpace 關聯的元素
        elements = space.ContainsElements
        for element in elements:
            # 如果元素有 IsContainedIn 屬性
            if hasattr(element, "IsContainedIn"):
                # 取得所屬的 IfcBuildingStorey
                building_storey = element.IsContainedIn[0].RelatingStructure
                break
    # 如果 IsDecomposedBy 屬性為空且 ContainsElements 屬性中也沒有元素，就跳過這個 IfcSpace

   
    # 取得 IfcBuildingStorey 的 GlobalId
building_storey_id = building_storey.GlobalId`

**Reply by Moult** [reply]
Yes those elevations are local Z values so you need to multiply by the object matrix. I've seen a few requests for top / bottom elevations from a few people so I've added it to the shape utils. See https://github.com/IfcOpenShell/IfcOpenShell/commit/32f5a77d0702e03af126236068db523ba2ac5db8

**Reply by u0416033** [reply]
@Moult Thanks a lot . Do I just need to replace my local shape.py of ifcopenshell package with these code ?

**Reply by u0416033** [reply]
I try it but I think the function is also output the SweptSolid I can not get the real real storey height of IFCSPACE. If it is in the 7F or 8F and it need to be diffirent height.

**Reply by aothms** [reply]
Placements are recursive. You can also simply do:

```python
import ifcopenshell.util.placement
ifcopenshell.util.placement.get_local_placement(space.ObjectPlacement)[2,3]
```

Without knowing for sure, I think this is the value provided by bimvision. So without taking into account the geometry. But taking into account the placement and all placements of the containing/aggregating elements.

**Reply by Moult** [reply]
@u0416033 you need to use the get shape top bottom elevation, the pure get elevation function is local coordinates.

**Reply by u0416033** [reply]
@aothms @Moult Thank you a lot .I have get the Placement of the IFCSPACE. I have a question about is it posible to get the IFCPOLYLINE of the IFCSPACE swepsolid .Because my target space have a hole inside , I try to get it like:

`    shape = ifcopenshell.geom.create_shape(settings, space)`
    
`    verts = shape.geometry.verts`
but not vertices I need polyline.

The suituiation situation like it below
![image](https://user-images.githubusercontent.com/53708536/225840764-2fe04356-1d1c-4e49-8195-3d6ab1437a66.png)
[https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/figures/ifcspace_2d-layout1.gif](url)

**Reply by u0416033** [reply]
I see the IFC file it has this information about the closed polyline of the IFCSPACE and it will be (#1336,#1346,#1356,#1374,#1384) . Is there anyway to get this five polyline vertices ?

#1391= IFCEXTRUDEDAREASOLID(#1386,#1390,#21,400.);
#1392= IFCSHAPEREPRESENTATION(#117,'Body','SweptSolid',(#1391));
#1394= IFCPRODUCTDEFINITIONSHAPE($,$,(#1392));

#1386= IFCARBITRARYPROFILEDEFWITHVOIDS(.AREA.,$,#1326,(#1336,#1346,#1356,#1374,#1384));
#1336= IFCPOLYLINE((#1328,#1330,#1332,#1334,#1328));

**Reply by aothms** [reply]
A somewhat reasonable approach would be:

[i for i in ifcfile.traverse(space.Representation) if i.is_a('IFCARBITRARYPROFILEDEFWITHVOIDS')][0].InnerCurves

i.e traverse over all instances attributes recursively, filter on this specific entity type and take the inner curves.

The downside is that it would be specific to an IFC entity type and you have to replicate the intelligence wrt to placements.

Another approach would be to use pythonocc. Use create_shape() with settings(USE_PYTHON_OPENCASCADE=True). Then use TopExp_Explorer to iterator over Face. Use TopoDS_Iterator to iterate over bounds. Use BRepTools::OuterWire() to check which of these wires is inner (i.e not outer). Use BRepTools_WireExplorer to then iterate over the verts of the wire.

The benefit is that you explore the geometry in world coords and works irrespective of the structure of the IFC. E.g the same inner bounds could be the result of a boolean operation.
